File: binding-recycler-view/src/main/java/dmitriy/khalturin/com/view/recycler/binding/adapter/BindingRecyclerAdapter.java
package dmitriy.khalturin.com.view.recycler.binding.adapter;

import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Dmitriy Khalturin <dmitry.halturin.86@gmail.com>
 * for android-binding-recycler-view on 14.04.18 19:13.
 */

public class BindingRecyclerAdapter<T> extends RecyclerView.Adapter<BindingViewHolder> {

  private int mLayoutId;
  private List<T> mItems;
  private int mVariableId;

  public BindingRecyclerAdapter(@LayoutRes int mLayoutId, @Nullable List<T> mItems, int mVariableId){
    this.mLayoutId = mLayoutId;
    this.mItems = (mItems != null ? mItems : new ArrayList<>());
    this.mVariableId = mVariableId;
  }

  @NonNull
  @Override
  public BindingViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType){
    View view = LayoutInflater.from(parent.getContext())
      .inflate(mLayoutId, parent, false);

    return new BindingViewHolder(view);
  }

  @Override
  public void onBindViewHolder(@NonNull BindingViewHolder holder, int position){
    T value = mItems.get(position);

    holder.getBinding().setVariable(mVariableId, value);
  }

  @Override
  public int getItemCount(){
    return mItems.size();
  }

  public List<T> getItems(){
    return mItems;
  }

  public void setItems(@Nullable List<T> mItems){
    if(mItems == null){
      this.mItems.clear();
    }else{
      this.mItems = mItems;
      notifyDataSetChanged();
    }
  }

}

File: library/src/main/kotlin/com/khalturin/dmitriy/library/recyclerview/adapter/BindingRecyclerAdapter.kt
package com.khalturin.dmitriy.library.recyclerview.adapter

import android.support.annotation.LayoutRes
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import io.reactivex.subjects.PublishSubject
import java.util.ArrayList

/**
 * Created by Dmitriy Khalturin <dmitry.halturin.86@gmail.com>
 * for android-rss-reader on 01.11.18 23:01.
 */
class BindingRecyclerAdapter<T>(
  @LayoutRes private val mLayoutId: Int,
  mInitItems: MutableList<T>?,
  private val mVariableId: Int
) : RecyclerView.Adapter<BindingViewHolder>() {

  private val mItems = mInitItems ?: ArrayList()

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = BindingViewHolder(
    LayoutInflater.from(parent.context).inflate(mLayoutId, parent, false)
  )

  override fun getItemCount() =  mItems.size

  override fun onBindViewHolder(holder: BindingViewHolder, position: Int) {
    holder.mBinding?.setVariable(mVariableId, mItems[position])
  }

  fun getItems() = mItems

  fun setItems(items: MutableList<T>?) {
    mItems.clear()
    mItems.addAll(items ?: ArrayList())

    notifyDataSetChanged()

    mSubjectItemsChanged.onNext(mItems.size)
  }

  private val mSubjectItemsChanged = PublishSubject.create<Int>()

  fun getOnItemsChanged() = mSubjectItemsChanged

}
